source/boundary_composition/function.cc

```diff
@@ -0,0 +1,219 @@
+          return numbers::signaling_nan<double>();
+        }
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::update()
+    {
+      // we get time passed as seconds (always) but may want
+      // to reinterpret it in years
+      if (this->convert_output_to_years())
+        boundary_composition_function.set_time (this->get_time() / year_in_seconds);
+      else
+        boundary_composition_function.set_time (this->get_time());
+    }
+
+
+
+    template <int dim>
+    double
+    Function<dim>::
+    minimal_composition (const std::set<types::boundary_id> &) const
+    {
+      return min_composition;
+    }
+
+
+
+    template <int dim>
+    double
+    Function<dim>::
+    maximal_composition (const std::set<types::boundary_id> &) const
+    {
+      return max_composition;
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Boundary composition model");
+      {
+        prm.enter_subsection("Function");
+        {
+          prm.declare_entry ("Coordinate system", "cartesian",
+                             Patterns::Selection ("cartesian|spherical|depth"),
+                             "A selection that determines the assumed coordinate "
+                             "system for the function variables. Allowed values "
+                             "are 'cartesian', 'spherical', and 'depth'. 'spherical' coordinates "
+                             "are interpreted as r,phi or r,phi,theta in 2D/3D "
+                             "respectively with theta being the polar angle. 'depth' "
+                             "will create a function, in which only the first "
+                             "parameter is non-zero, which is interpreted to "
+                             "be the depth of the point.");
+
+          Functions::ParsedFunction<dim>::declare_parameters (prm, 1);
+
+          prm.declare_entry ("Minimal composition", "0",
+                             Patterns::Double (),
+                             "Minimal composition. Units: none.");
+          prm.declare_entry ("Maximal composition", "1",
+                             Patterns::Double (),
+                             "Maximal composition. Units: none.");
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      // we need to get at the number of compositional fields here to
+      // initialize the function parser. unfortunately, we can't get it
+      // via SimulatorAccess from the simulator itself because at the
+      // current point the SimulatorAccess hasn't been initialized
+      // yet. so get it from the parameter file directly.
+      prm.enter_subsection ("Compositional fields");
+      const unsigned int n_compositional_fields = prm.get_integer ("Number of fields");
+      prm.leave_subsection ();
```

Comment:
Replace this the same way you did in your other pull request. Also remove the outdated comment above.